mlflow/transformers.py

```diff
@@ -990,3 +1110,291 @@ def from_dict(
+        if isinstance(output_data, list):
+            output_coll = []
+            for output in output_data:
+                if isinstance(output, dict):
+                    output_coll.append(output[target_dict_key])
+                elif isinstance(output, list):
+                    output_coll.append(
+                        self._parse_lists_of_dict_to_list_of_str(output, target_dict_key)[0]
+                    )
+            return output_coll
+        else:
+            return output_data[target_dict_key]
+
+    def _parse_tokenizer_output(self, output_data, target_dict_key):
+        # NB: We're collapsing the results here to a comma separated string for each inference
+        # input string. This is to simplify having to otherwise make extensive changes to
+        # ColSpec in order to support schema enforcement of List[List[str]]
+        if isinstance(output_data[0], list):
+            return [self._parse_tokenizer_output(coll, target_dict_key) for coll in output_data]
+        else:
+            return ",".join([coll[target_dict_key] for coll in output_data])
+
+    @staticmethod
+    def _parse_list_of_multiple_dicts(output_data, target_dict_key):
+        if isinstance(output_data[0], list):
+            return [collection[0][target_dict_key] for collection in output_data]
+        else:
+            return [output_data[0][target_dict_key]]
+
+    def _parse_list_output_for_multiple_candidate_pipelines(self, output_data):
+        # NB: This will not continue to parse nested lists. Pipelines do not output complex
+        # types that are greater than 2 levels deep so there is no need for more complex
+        # traversal for outputs.
+        if isinstance(output_data[0], list):
+            return [
+                self._parse_list_output_for_multiple_candidate_pipelines(x) for x in output_data
+            ]
+        else:
+            return output_data[0]
+
+    def _parse_question_answer_input(self, data):
+        """
+        Parses the single string input representation for a question answer pipeline into the
+        required dict format for a `question-answering` pipeline.
+        """
+        if isinstance(data, str):
+            if any(key not in data for key in ["question: ", "context: "]):
+                raise MlflowException(
+                    "There is an error in the syntax of the provided string. A "
+                    "question and a context must be provided in the format: "
+                    "`question: <text> context: <text>`. Validate that the "
+                    "required keys are provided."
+                )
+            else:
+                first, last = data.split("context:")
+                if first:
+                    return {
+                        "question": first.split("question:")[1].strip(),
+                        "context": last.strip(),
+                    }
+                else:
+                    context, question = last.split("question:")
```

Comment:
```suggestion
                    context, question = last.split("question:", 1)
```

Reply:
I refactored this to use Dict inputs since the string parsing was very confusing and potentially error-prone.